runtime/lua/vim/treesitter/query.lua

```diff
@@ -179,11 +179,19 @@ function M.parse_query(lang, query)
   end
 end
 
---- Gets the text corresponding to a given node
+--- Gets the text corresponding to a given node or a list of nodes
 ---
----@param node the node
+---@param node the node or list of nodes
 ---@param source The buffer or string from which the node is extracted
 function M.get_node_text(node, source)
+  -- If node is a list of nodes then return their concatenated text
+  if node['start'] == nil and type(node) == 'table' then
+    local text = ''
+    for _, node1 in ipairs(node) do
+      text = text .. M.get_node_text(node1, source)
```

Comment:
string concatenation in Lua is horribly slow. A better idiom is to populate a table first and then call `table.concat`.

However, it may be better to change the signature of this to return a list of strings?

Reply:
I implemented it this way, so `get_node_text(match[capture_id])`, which works with the old behavior, still gives the same result for queries that only have one node per capture. A lot of code I saw uses `iter_matches` in that way, so maybe it would make sense to keep compatibility in this way?

Reply:
I meant the return type of course. The `table.concat` solution would definitely be an improvement.

Reply:
Sure. When passing a single node which isn't a list, a single string should be returned.

I'm just contemplating whether if a list is given as the input, then a list is returned? Should remain compat and doesn't remove information.

Reply:
I'm thinking of this specific use case `get_node_text(match[capture_id])`, which I saw in quite a bit of code, e.g. most of the currently implemented query predicates.

If `get_node_text` were to return a list of strings, the type of that expression would change to "list of string". If a single string is returned, then the type of that expression stays "string" (even though the type of `match[capture_id]` changed).

If someone needs the additional information that a list of string has, they will probably want to iterate over that list next, in which case it would be easier (and faster, as this doesn't build the extra table) to just iterate over the list of nodes and then call `get_node_text` in the loop.

Reply:
related: https://github.com/neovim/neovim/pull/18232

Reply:
> Sure. When passing a single node which isn't a list, a single string should be returned.
> 
> I'm just contemplating whether if a list is given as the input, then a list is returned? Should remain compat and doesn't remove information.

I disagree; return types should be static. Either we need two functions, or (as an uneasy compromise)  add an `opts` table that controls the behavior.